File: ios/Src/ReactNative/RCTPopoverView/RCTPopoverViewManager.m
//
//  RCTPopoverViewManager.m
//  IosPopoverExample
//
//  Created by Dominic Go on 11/27/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

#import <React/RCTViewManager.h>
 

@interface RCT_EXTERN_MODULE(RCTPopoverViewManager, RCTViewManager)

// ---------------------------
// MARK: Export Props - Events
// ---------------------------

RCT_EXPORT_VIEW_PROPERTY(onPopoverWillShow, RCTBubblingEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onPopoverWillHide, RCTBubblingEventBlock);

RCT_EXPORT_VIEW_PROPERTY(onPopoverDidShow, RCTBubblingEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onPopoverDidHide, RCTBubblingEventBlock);

// ---------------------------
// MARK: Export Props - Values
// ---------------------------

RCT_EXPORT_VIEW_PROPERTY(popoverSize, NSString);
RCT_EXPORT_VIEW_PROPERTY(permittedArrowDirections, NSArray);
RCT_EXPORT_VIEW_PROPERTY(popoverBackgroundColor, NSNumber);

// --------------------------
// MARK: Export Props - Flags
// --------------------------

RCT_EXPORT_VIEW_PROPERTY(popoverShouldDismiss, BOOL);
RCT_EXPORT_VIEW_PROPERTY(popoverCanOverlapSourceViewRect, BOOL);


// ---------------------------
// MARK: View Manager Commands
// ---------------------------

RCT_EXTERN_METHOD(setVisibility: (nonnull NSNumber *)node
                  visibility   : (nonnull BOOL     *)visibility);

@end

File: ios/Src/ReactNative/RCTPopoverView/RCTPopoverViewManager.swift
//
//  RCTPopoverViewManager.swift
//  IosPopoverExample
//
//  Created by Dominic Go on 11/27/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

import Foundation

@objc(RCTPopoverViewManager)
class RCTPopoverViewManager: RCTViewManager {
  
  static var sharedBridge: RCTBridge? {
    didSet {
      #if DEBUG
      print("RCTPopoverView, sharedBridge: didSet");
      NotificationCenter.default.addObserver(RCTPopoverViewManager.self,
        selector: #selector(Self.resetSharedBridge),
        name: NSNotification.Name(rawValue: "RCTBridgeWillReloadNotification"),
        object: nil
      );
      #endif
    }
  };
  
  /// invalidate RCTBridge instance
  @objc static func resetSharedBridge() {
    #if DEBUG
    print("RCTPopoverView: resetSharedBridge...");
    #endif
    Self.sharedBridge = nil;
  };
    
  override static func requiresMainQueueSetup() -> Bool {
    return true;
  };
  
  override func view() -> UIView! {
    if Self.sharedBridge == nil {
      Self.sharedBridge = self.bridge;
    };
    
    return RCTPopoverView(bridge: self.bridge);
  };
  
  @objc func setVisibility(_ node: NSNumber, visibility: Bool){
    DispatchQueue.main.async {
      guard let view = self.bridge.uiManager.view(forReactTag: node),
            let popoverView = view as? RCTPopoverView
      else { return };
      
      popoverView.setVisibility(visibility);
    };
  };
};
